File: src/api/refinement.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from fastapi.responses import PlainTextResponse
from sqlalchemy.orm import Session

from src.api.schemas_refinement import (
    CreateSessionRequest,
    RenameSessionRequest,
    SessionDetailResponse,
    SessionListItem,
    SessionListResponse,
    SetModeRequest,
    StartSessionResponse,
    SubmitAnswersRequest,
    SubmitAnswersResponse,
)
from src.database import get_db
from src.services.refinement_service import get_refinement_service

router = APIRouter(prefix="/api/refinement", tags=["refinement-api"])


@router.post("/sessions", response_model=StartSessionResponse)
async def create_session(payload: CreateSessionRequest, db: Session = Depends(get_db)):
    service = get_refinement_service()
    try:
        return await service.start_session(db, payload)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/sessions", response_model=SessionListResponse)
async def list_sessions(
    q: str | None = None,
    status: str | None = None,
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
):
    service = get_refinement_service()
    return service.list_sessions(db, query=q, status=status, limit=limit, offset=offset)


@router.get("/sessions/{session_id}", response_model=SessionDetailResponse)
async def get_session(session_id: str, db: Session = Depends(get_db)):
    service = get_refinement_service()
    try:
        return await service.get_session(db, session_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.patch("/sessions/{session_id}", response_model=SessionListItem)
async def rename_session(
    session_id: str,
    payload: RenameSessionRequest,
    db: Session = Depends(get_db),
):
    service = get_refinement_service()
    try:
        return service.rename_session(db, session_id, payload.title)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.delete("/sessions/{session_id}", status_code=204)
async def delete_session(session_id: str, db: Session = Depends(get_db)):
    service = get_refinement_service()
    try:
        service.delete_session(db, session_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return Response(status_code=204)


@router.post("/sessions/{session_id}/mode", response_model=SessionDetailResponse)
async def set_session_mode(session_id: str, payload: SetModeRequest, db: Session = Depends(get_db)):
    service = get_refinement_service()
    try:
        return await service.set_mode(db, session_id, payload.mode)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/sessions/{session_id}/answers", response_model=SubmitAnswersResponse)
async def submit_answers(
    session_id: str,
    payload: SubmitAnswersRequest,
    db: Session = Depends(get_db),
):
    service = get_refinement_service()
    try:
        return await service.submit_answers(db, session_id, payload)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/sessions/{session_id}/export", response_class=PlainTextResponse)
async def export_markdown(session_id: str, db: Session = Depends(get_db)):
    service = get_refinement_service()
    try:
        markdown = await service.export_markdown(db, session_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return PlainTextResponse(
        markdown,
        headers={"Content-Disposition": f'attachment; filename="refinement-{session_id}.md"'},
    )

File: frontend/src/api/refinement.ts
import { apiFetch } from "./client";
import type {
  CreateSessionRequest,
  ListSessionsParams,
  RenameSessionRequest,
  SessionDetailResponse,
  SessionListItem,
  SessionListResponse,
  SetModeRequest,
  StartSessionResponse,
  SubmitAnswersRequest,
  SubmitAnswersResponse,
} from "../types/api";

const BASE = "/api/refinement";

export function createSession(payload: CreateSessionRequest): Promise<StartSessionResponse> {
  return apiFetch(`${BASE}/sessions`, { method: "POST", body: JSON.stringify(payload) });
}

export function listSessions(params: ListSessionsParams = {}): Promise<SessionListResponse> {
  const search = new URLSearchParams();
  if (params.q?.trim()) search.set("q", params.q.trim());
  if (params.status) search.set("status", params.status);
  if (params.limit !== undefined) search.set("limit", String(params.limit));
  if (params.offset !== undefined) search.set("offset", String(params.offset));
  const query = search.toString();
  return apiFetch(`${BASE}/sessions${query ? `?${query}` : ""}`);
}

export function renameSession(
  sessionId: string,
  payload: RenameSessionRequest,
): Promise<SessionListItem> {
  return apiFetch(`${BASE}/sessions/${encodeURIComponent(sessionId)}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteSession(sessionId: string): Promise<void> {
  return apiFetch(`${BASE}/sessions/${encodeURIComponent(sessionId)}`, { method: "DELETE" });
}

export function getSession(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch(`${BASE}/sessions/${encodeURIComponent(sessionId)}`);
}

export function setSessionMode(
  sessionId: string,
  payload: SetModeRequest,
): Promise<SessionDetailResponse> {
  return apiFetch(`${BASE}/sessions/${encodeURIComponent(sessionId)}/mode`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function submitAnswers(
  sessionId: string,
  payload: SubmitAnswersRequest,
): Promise<SubmitAnswersResponse> {
  return apiFetch(`${BASE}/sessions/${encodeURIComponent(sessionId)}/answers`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function exportUrl(sessionId: string): string {
  return `${BASE}/sessions/${encodeURIComponent(sessionId)}/export`;
}
